lib/scholar/linear/polynomial_regression.ex

```diff
@@ -0,0 +1,190 @@
+      %Scholar.Linear.LinearRegression{
+        coefficients: #Nx.Tensor<
+          f32[5]
+          [-0.021396497264504433, -0.004854593891650438, -0.08849877119064331, -0.062211357057094574, -0.04369127377867699]
+        >, intercept: #Nx.Tensor<
+          f32
+          4.418517112731934
+        >
+      }
+  """
+  deftransform fit(a, b, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    a_transform = transform(a, opts |> Keyword.put(:fit_intercept?, false))
+    Scholar.Linear.LinearRegression.fit(a_transform, b, Keyword.delete(opts, :degree))
+  end
+
+  @doc """
+  Computes the feature matrix for polynomial regression.
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Examples
+
+      iex> x = Nx.tensor([[2]])
+      iex> Scholar.Linear.PolynomialRegression.transform(x, degree: 0)
+      ** (NimbleOptions.ValidationError) invalid value for :degree option: expected positive integer, got: 0
+
+      iex> x = Nx.tensor([[2]])
+      iex> Scholar.Linear.PolynomialRegression.transform(x, degree: 5, fit_intercept?: false)
+      #Nx.Tensor<
+        s64[1][5]
+        [
+          [2, 4, 8, 16, 32]
+        ]
+      >
+
+      iex> x = Nx.tensor([[2, 3]])
+      iex> Scholar.Linear.PolynomialRegression.transform(x)
+      #Nx.Tensor<
+        s64[1][6]
+        [
+          [1, 2, 3, 4, 6, 9]
+        ]
+      >
+
+      iex> x = Nx.iota({3, 2})
+      iex> Scholar.Linear.PolynomialRegression.transform(x, fit_intercept?: false)
+      #Nx.Tensor<
+        s64[3][5]
+        [
+          [0, 1, 0, 0, 1],
+          [2, 3, 4, 6, 9],
+          [4, 5, 16, 20, 25]
+        ]
+      >
+  """
+  deftransform transform(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    transform_n(x, opts)
+  end
+
+  deftransform transform_n(x, opts) do
+    x_split = initial_data_split(x)
+
+    1..(opts[:degree] - 1)//1
+    |> Enum.reduce([x_split], fn _, prev_degree ->
+      [prev_degree, compute_degree(x, List.last(prev_degree))]
+    end)
+    |> List.flatten()
+    |> Nx.concatenate(axis: 1)
+    |> add_intercept(opts)
+  end
+
+  @spec initial_data_split(Nx.Tensor) :: list(Nx.Tensor)
+  deftransform initial_data_split(x) do
+    {n_samples, n_features} = Nx.shape(x)
+
+    Enum.map(
+      0..(n_features - 1),
+      fn n ->
+        Nx.reshape(x[[0..-1//1, n]], {n_samples, :auto})
+      end
+    )
+  end
+
+  @spec compute_degree(Nx.Tensor, list(Nx.Tensor)) :: list(Nx.Tensor)
+  deftransform compute_degree(x, previous_degree) do
+    {_n_samples, n_features} = Nx.shape(x)
+
+    Enum.map(0..(n_features - 1), fn nf ->
+      previous_degree
+      |> Enum.slice(nf..-1)
+      |> Nx.concatenate(axis: 1)
+      |> compute_column(x, nf)
+    end)
+  end
+
+  defnp compute_column(previous, x, n) do
+    {n_samples, _n_features} = Nx.shape(x)
+
+    x[[0..-1//1, n]]
+    |> Nx.reshape({n_samples, :auto})
+    |> Nx.multiply(previous)
```

Comment:
multiplying by some sort of previous value sounds like we could apply Nx.cumulative_product somehow for this scaling, but I'm not sure about that

Reply:
perhaps finding a reference implementation for this in scipy, jax or sklearn would be helpful

Reply:
by this I mean the whole calculation, not just this `compute_degree` function

Reply:
> multiplying by some sort of previous value sounds like we could apply Nx.cumulative_product somehow for this scaling, but I'm not sure about that

"previous" makes reference to the previous degree iteration of the operation, so the columns that are multiplied are not (usually) adjacent